applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+    no adjoint solution is necessary. except from that it works the same as
+    'AdjointResponseFunction'.
+
+    Attributes
+    ----------
+    primal_analysis : Primal analysis object of the response function
+    response_function_utility: Cpp utilities object doing the actual computation of response value and gradient.
+    """
+    def __init__(self, identifier, project_parameters, model_part = None):
+        self.identifier = identifier
+        self.response_function_utility = StructuralMechanicsApplication.AdjointStrainEnergyResponseFunction( model_part, project_parameters )
+
+        model = Model()
+        model.AddModelPart(model_part)
+        self.primal_model_part = model_part
+
+        # Create the primal solver
+        self.ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, self.ProjectParametersPrimal)
+
+        self.primal_model_part.AddNodalSolutionStepVariable(SHAPE_SENSITIVITY)
+
+        # Add variables to save the solution of the adjoint problem
+        self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT)
+        if self.ProjectParametersPrimal["solver_settings"]["rotation_dofs"].GetBool():
+            self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_ROTATION)
+        #self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.POINT_LOAD_SENSITIVITY)
+        # TODO: Is it necessary to add other variables (e.g. POINT_LOAD_SENSITIVITY)?
+
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.IS_ADJOINT] = False
+
+    def Initialize(self):
+        self.primal_analysis.Initialize()
+
+    def InitializeSolutionStep(self):
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
+        self.primal_analysis._GetSolver().SolveSolutionStep()
+        self.primal_analysis.FinalizeSolutionStep()
+        self.primal_analysis.OutputSolutionStep()
+        print("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
+
+    def CalculateValue(self):
+        startTime = timer.time()
+        value = self.response_function_utility.CalculateValue(self.primal_model_part)
+        print("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
+
+    def CalculateGradient(self):
+        # Replace elements and conditions by its adjoint equivalents
+        self.__PerformReplacementProcess()
```

Comment:
Why do we need to swap primal and adjoint elements here?

Reply:
This was explained in person. Due to the large number of peculiarities in this response function, I would be in favor of replacing it with a simpler, more generic albeit more computationally expensive version which follows the others and works for a larger class of problems, but I will not push this here.

Reply:
The adjoint linear strain energy response is now also using the adjoint solver - no need for this special treatment anymore